### ISSUE TITLE
Add a manylinux2014 compatible release with only LLVM-based backends

### ISSUE THREAD

**Issue by k-ye** [taichi-dev/taichi#4395]
This will include both CPU and CUDA backends, which could be useful for cloud users. See https://github.com/taichi-dev/taichi/issues/4377, https://github.com/taichi-dev/taichi/issues/3332

**Comment by bobcao3** [comment]
If we directly build on CentOS 7 nvidia docker, can we get a manylinux2014 wheel with proper cuda / vulkan?

**Comment by qiao-bo** [comment]
@k-ye @bobcao3, Let me recap some of our previous efforts here: The problem with CentOS 7 is that it comes with GCC 4.8, this is too old to compile Taichi (not saying impossible, but too many errors to dig). This is the same story for Nvidia CentOS7. The way we circumvent this is to use a magic docker image with Red Hat devtoolset-10. (https://git.centos.org/rpms/devtoolset-10-gcc), which is based on CentOS7 but with gcc 10. By using this image (quay.io/pypa/manylinux2014_x86_64), we still need to recompile clang and llvm but it is possible to build. 

Now to support CUDA, I see three options:
[a] Use Nvidia CentOS 7 image, but fix gcc 4.8 problem. Difficulty is high.
[b] Use the CentOS 7 based image with gcc 10. But install and build Nvidia/CUDA tools. Should be possible.
[c] Still use our Ubuntu 18.04 image, but pin glibc version to 2.2.5 (we already knew how to do it). But ship Taichi without Vulkan. Difficulty is low.

I vote for option [c], WDYT?

**Comment by k-ye** [comment]
[c] sounds good to me!

**Comment by bobcao3** [comment]
c sounds like the best option but I do think we should attempt `a` or `b` so that at least we can confirm vulkan works on the manylinux2014 target. (Which is a lot higher than 2.2.5)

**Comment by k-ye** [comment]
I think our top priority *in this issue* is to unblock CUDA users.. Vulkan will be a nice-to-have, but it's not that urgent :-)

**Comment by strongoier** [comment]
Where can we upload such a wheel without Vulkan?

**Comment by k-ye** [comment]
> Where can we upload such a wheel without Vulkan?

I believe GH provides API to upload additional artifacts to release pages like this (https://github.com/taichi-dev/taichi/releases/tag/v0.9.0)

**Comment by strongoier** [comment]
Yes that's definitely possible. However, the updated version is no longer Taichi v9.0. Are we going to do so only for future releases?

**Comment by k-ye** [comment]
> Are we going to do so only for future releases?

Yep

>  to manually update that wheel to the GH release page

I think this can be automated with actions like  https://github.com/marketplace/actions/gh-release

**Comment by qiao-bo** [comment]
GCC 4.8 comes with [CXXABI 1.3.7](https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html). Maybe first try if the wheel works with this image (quay.io/pypa/manylinux2014_x86_64)? If yes, option [b] will probably have the same result.

**Comment by strongoier** [comment]
Same error appears on `quay.io/pypa/manylinux2014_x86_64`. The magic image works by updating GCC without updating `libstdc++`, so we can still only use `CXXABI 1.3.7`.

**Comment by qiao-bo** [comment]
[This dockerfile](https://github.com/taichi-dev/taichi/blob/master/ci/Dockerfile.manylinux2014.cpu) builds Taichi on CPU using the image. It is useful as a starting point.

**Comment by strongoier** [comment]
I finally choose path [a], customizing an image for our purpose. After tons of trials, now `ti.cuda` works properly and:
```
$ auditwheel-symbols -m 2014 taichi-0.9.2-cp39-cp39-linux_x86_64.whl 
taichi/_lib/core/taichi_core.so is manylinux_2_17(aka manylinux2014) compliant.
```

Next steps:
- [ ] Upload the Dockerfile.
- [ ] Add the image to our local registry.
- [ ] Add the generation of the wheel into the release workflow.
- [ ] Add an auto-uploading step in the auto-release script.
